### PR TITLE
Add support for adding/removing a passive

### PR DESCRIPTION
**Pull request by ramsai1729**

### CODE REVIEW

**Comment by ramsai1729** [comment]
@myronkscott could you review this PR?

File: dso-l2/src/main/java/com/tc/objectserver/impl/TopologyManager.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  private volatile TopologyMbean topologyMbean;
+  private final List<TopologyListener> listeners = new ArrayList<>();
+
+  private TopologyManager() {
+  }
+
+  private static final TopologyManager INSTANCE = new TopologyManager();
+
+  public static TopologyManager get() {
+    return INSTANCE;
```
Comment (myronkscott) [inline dso-l2/src/main/java/com/tc/objectserver/impl/TopologyManager.java]: Can this be done without singletons?

File: dso-l2/src/main/java/com/tc/l2/state/ConsistencyManager.java [hunk 2]
```diff
@@ -53,7 +55,11 @@ boolean isStateTransition() {
     }
   }
 
-  boolean requestTransition(ServerMode mode, NodeID sourceNode, Transition newMode) throws IllegalStateException;
+  default boolean requestTransition(ServerMode mode, NodeID sourceNode, Transition newMode) throws IllegalStateException {
+    return requestTransition(mode, sourceNode, TopologyManager.get().getTopology(), newMode);
```
Comment (myronkscott) [inline dso-l2/src/main/java/com/tc/l2/state/ConsistencyManager.java]: maybe get rid of the singleton and pass null?  use legacy behavior when null is passed.

File: dso-l2/src/main/java/com/tc/l2/state/ConsistencyManagerImpl.java [hunk 3]
```diff
@@ -132,25 +135,26 @@ public boolean requestTransition(ServerMode mode, NodeID sourceNode, Transition
     boolean allow = false;
     
     // activate voting to lock the voting members and return the number of server votes
-    int serverVotes = activateVoting(mode, newMode);
+    int serverVotes = activateVoting(mode, newMode, topology);
+    int peerServers = topology.getServers().size() - 1;
```
Comment (myronkscott) [inline dso-l2/src/main/java/com/tc/l2/state/ConsistencyManagerImpl.java]: topology is locked while voting is occurring correct?  How about filtering the list of active peers before voting starts and using that for endVoting?  I'm concerned that it is not clear here that topology will not change during a vote.

File: dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java [hunk 4]
```diff
@@ -249,16 +252,16 @@ private synchronized void electionStarted(Enrollment e, State serverState) {
...
   }
 
-  private NodeID doElection(NodeID myNodeId, boolean isNew, WeightGeneratorFactory weightsFactory, State currentState)
+  private NodeID doElection(NodeID myNodeId, Topology topology, boolean isNew, WeightGeneratorFactory weightsFactory, State currentState)
       throws GroupException, InterruptedException {
 
     // Step 1: publish to cluster NodeID, weight and election start
     Enrollment e = EnrollmentFactory.createEnrollment(myNodeId, isNew, weightsFactory);
-    electionStarted(e, currentState);
+    electionStarted(e, currentState, topology);
```
Comment (myronkscott) [inline dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java]: let's lock down the topology here by getting the servers now and passing that to election started.

File: dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java [hunk 5]
```diff
@@ -249,16 +252,16 @@ private synchronized void electionStarted(Enrollment e, State serverState) {
...
 
     L2StateMessage msg = L2StateMessage.createElectionStartedMessage(e, currentState);
     debugInfo("Sending my election vote to all members");
-    groupManager.sendAll(msg);
+    groupManager.sendTo(topology.getServers(), msg);
```
Comment (myronkscott) [inline dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java]: use the local server list that you will generate at the beginning here.

File: dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java [hunk 6]
```diff
@@ -274,7 +277,7 @@ private NodeID doElection(NodeID myNodeId, boolean isNew, WeightGeneratorFactory
     // Step 4 : local host won the election, so notify world for acceptance
     msg = L2StateMessage.createElectionResultMessage(e, currentState);
     debugInfo("Won election, announcing to world and waiting for response...");
-    GroupResponse<L2StateMessage> responses = groupManager.sendAllAndWaitForResponse(msg);
+    GroupResponse<L2StateMessage> responses = groupManager.sendToAndWaitForResponse(topology.getServers(), msg);
```
Comment (myronkscott) [inline dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java]: use the local server list that you will generate at the beginning here.

File: dso-l2/src/main/java/com/tc/l2/state/StateManagerImpl.java [hunk 7]
```diff
@@ -200,19 +203,23 @@ private void runElection() {
     if (!electionStarted()) {
       return;
     }
+
+    // This topology will be used throughout the election process
+    Topology topology = TopologyManager.get().getTopology();
+
```
Comment (myronkscott) [inline dso-l2/src/main/java/com/tc/l2/state/StateManagerImpl.java]: can this be passed in from somewhere.  Singletons don't look right here.

File: dso-l2/src/main/java/com/tc/net/groups/Node.java [hunk 8]
```diff
@@ -19,23 +19,33 @@
...
+
   private final String host;
   private final int    port;
   private final int    groupPort;
   private final int    hashCode;
 
+  public Node(String hostPort) {
+    this(getHost(hostPort), getPort(hostPort));
+  }
+
```
Comment (myronkscott) [inline dso-l2/src/main/java/com/tc/net/groups/Node.java]: are you just adding this for code reuse convenience in passing host:port?

File: dso-l2/src/main/java/com/tc/net/groups/TCGroupManagerImpl.java [hunk 9]
```diff
@@ -515,6 +520,11 @@ public void sendTo(NodeID node, AbstractGroupMessage msg) throws GroupException
     internalSendTo(node, msg, sentCallback);
   }
 
+  @Override
+  public void sendTo(Set<String> nodes, AbstractGroupMessage msg) {
+    sendAll(msg, members.keySet().stream().filter(id -> nodes.contains(id.getName())).collect(toSet()));
+  }
+
```
Comment (myronkscott) [inline dso-l2/src/main/java/com/tc/net/groups/TCGroupManagerImpl.java]: Why are these new filtered sendAll's being added?  is there lag in updating the member set?

File: dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java (hunk 4 above)
Comment (ramsai1729) [reply dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java]: Yes, topology is already locked in `StateManagerImpl.java#runElection()` and passed here in `ElectionContext`

File: dso-l2/src/main/java/com/tc/net/groups/TCGroupManagerImpl.java (hunk 9 above)
Comment (ramsai1729) [reply dso-l2/src/main/java/com/tc/net/groups/TCGroupManagerImpl.java]: This change helps in sending election messages to locked members as topology might have changed while the election is running and a new server got added to `members`.

File: dso-l2/src/main/java/com/tc/objectserver/impl/TopologyManager.java (hunk 1 above)
Comment (ramsai1729) [reply dso-l2/src/main/java/com/tc/objectserver/impl/TopologyManager.java]: Yes but TopologyManager needs to be passed to all components where it is used. will try that.

File: dso-l2/src/main/java/com/tc/l2/state/StateManagerImpl.java (hunk 7 above)
Comment (ramsai1729) [reply dso-l2/src/main/java/com/tc/l2/state/StateManagerImpl.java]: changed now

File: dso-l2/src/main/java/com/tc/net/groups/Node.java (hunk 8 above)
Comment (ramsai1729) [reply dso-l2/src/main/java/com/tc/net/groups/Node.java]: yes

File: dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java (hunk 5 above)
Comment (ramsai1729) [reply dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java]: Separate local server list won't be necessary as Topology class is immutable and we don't need to change much when voter count changes supported.

File: dso-l2/src/main/java/com/tc/l2/state/ConsistencyManager.java (hunk 2 above)
Comment (ramsai1729) [reply dso-l2/src/main/java/com/tc/l2/state/ConsistencyManager.java]: done

**Comment by ramsai1729** [comment]
@myronkscott pushed some changes to address your comments. Please have a look.

File: dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java [hunk 10]
```diff
@@ -237,9 +239,10 @@ private synchronized void sendToNewMember(NodeID node) {
     }
   }
 
-  private synchronized void electionStarted(Enrollment e, State serverState) {
+  private synchronized void electionStarted(Enrollment e, State serverState, Topology topology) {
     if (this.state == ELECTION_IN_PROGRESS) { throw new AssertionError("Election Already in Progress"); }
```
[the remark this answers is not in the document]
Comment (ramsai1729) [reply dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java]: done

File: dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java [hunk 11]
```diff
@@ -249,16 +252,16 @@ private synchronized void electionStarted(Enrollment e, State serverState) {
     logger.info("Election Started : " + e);
   }
 
-  private NodeID doElection(NodeID myNodeId, boolean isNew, WeightGeneratorFactory weightsFactory, State currentState)
+  private NodeID doElection(NodeID myNodeId, Topology topology, boolean isNew, WeightGeneratorFactory weightsFactory, State currentState)
       throws GroupException, InterruptedException {
```
[the remark this answers is not in the document]
Comment (ramsai1729) [reply dso-l2/src/main/java/com/tc/l2/state/ElectionManagerImpl.java]: done

**Pull request merged**